clang/test/OpenMP/nothing_messages.cpp

```diff
@@ -12,7 +12,6 @@ int mixed() {
     x=d;
   }
 
-// expected-error@+2 {{#pragma omp nothing' cannot be an immediate substatement}}
```

Comment:
But I don't see such a restriction for `nothing` in spec 5.1, is it really required ?

Reply:
3 Directive and Construct Syntax
Neither a stand-alone directive nor a declarative directive may be used in place of a substatement
15 in a selection statement or iteration statement, or in place of the statement that follows a label.

Reply:
Seems to me, this should be allowed

Reply:
@alexey-bataev That restriction shouldn't apply to the `nothing` directive.  The `nothing` directive is not executable, and so is not considered a "stand-alone directive". Note that allowing the insertion of the `barrier` directive in the following would push the call to `f()` outside of the if block; that's why it is restricted:

```
if (x)
   #pragma omp barrier
   f();
```

On the other hand, allowing the `nothing` directive in the following should have no effect on the behavior of the program and therefore should be permitted:

```
if (x)
    #pragma omp nothing
   f();
```

Reply:
@dreachem The problem is, that the compiler has no idea what to do with this code.
`nothing` has no associated structured block, right? If so, then this code is actually this:
```
if (x)
    #pragma omp nothing
f();
```
With OpenMP support, this code represents something like this:
```
if (x)
  ;
f();
```
If we disable OpenMP support, this code will be transformed into this:
```
if (x)
  f();
```
because the directive will be ignored completely.
So, having something like this:
```
if (x)
  #pragma omp nothing
```
should not be allowed, it is not correct program if OpenMP is disabled.
To make it correct, you need to have something like this:
```
if (x) {
  #pragma omp nothing
}
```
This will be correct.
The check, you're trying to remove, checks exactly for such stand-alone directives, which may cause troubles when OpenMP is disabled.

Reply:
As I understand it, the `nothing` directive is not the equivalent of a null statement. It is an "ignore me" directive.

With OpenMP support, this:

```
if (x)
    #pragma omp nothing
f();
```

should be transformed to this (i.e., the nothing directive should simply be ignored):

```
if (x)
f();
```

If we are transforming it into a null statement, that is wrong.

Reply:
Ah, ok, I see. Then yes, this check must be removed.